MCR-3126 OCFL filesystem

Mewel opened the pull request:
[Link to jira](https://mycore.atlassian.net/browse/MCR-3126).

Mewel committed:
MCR-3126 ocfl draft

Mewel committed:
MCR-3126 update OCFL

* update ocfl-java version
* support file size
* support filekey

Mewel committed:
MCR-3126 improve ocfl file system

* add range request support
* fix codacy issues
* rm unecessary code

Mewel committed:
MCR-3126 improve ocfl

* fix truncate existing for remote repository
* fix isAdded() for root path
* add javadoc & junit tests

Mewel committed:
MCR-3126 improve ocfl

- add default ocfl properties
- add some javadoc

Mewel committed:
use mcrderivate: prefix for derivate objectId

Mewel committed:
abstract ocfl test cases; use files/ for ocfl content

Mewel committed:
force/check path if valid for the current virtual object

Mewel committed:
MCR-3126 fix pom.xml order

Mewel committed:
MCR-3126 improve ocfl

* unify ocfl version handling
* rm transaction folder on startup
* add javadoc and more test cases

Mewel committed:
MCR-3126 make ocfl compatible with changes in main

Mewel committed:
MCR-3126 fix javadoc

Mewel committed:
MCR-3126 fix license

Mewel committed:
MCR-3126 prepare for S3 implementation

Mewel committed:
MCR-3126 add s3 support

-use own RepositoryBuilder
-add S3RepositoryProvider
-improve directory history method by using inventory API

Mewel committed:
MCR-3126 sort pom.xml

Mewel committed:
MCR-3126 update pom.xml; revert MCRBase; fix PMD

Mewel committed:
MCR-3126 update dependencies

* update to ocfl-java to 2.2.2
* update aws sdk
* extend repository builder correctly

Mewel committed:
MCR-3126 fix pmd issues

Mewel committed:
MCR-3126 fix toPhysicalPath() method

Mewel committed:
MCR-3126 fix pmd rules

Mewel committed:
MCR-3126 correctly initialize rolling cache

Mewel committed:
MCR-3126 try to reduce usage of toPhysicalPath

Mewel committed:
MCR-3126 fix aws pom problems

Mewel committed:
MCR-3126 introduce remote filestore for OCFL

Mewel committed:
MCR-3126 add 'soft' delete support

Mewel committed:
MCR-3126 rm unecessary remote file store

Mewel committed:
MCR-3126 improve file tracker with better rename support

Mewel committed:
MCR-3126 fix bug because of API changes

Mewel committed:
MCR-3126 improve seekable chaching channel

Mewel committed:
MCR-3126 fix import order

Mewel committed:
MCR-3126 use junit5

Mewel committed:
MCR-3126 rm junit5 vintage engine & fix assertThrows

Mewel committed:
MCR-3126 fix pom.xml order

Mewel committed:
MCR-3126 fix codacy issues

Mewel committed:
simplify fileKey and probeContent

sebhofmann commented:
I found a conceptional Problem, which leads to errors in the tiling process. The MCRTreeCopier commits and restarts database transactions. The tiling queue runs in another thread is now able to find the commited tiling jobs and now starts tiling. The ocfl transaction of the Thread in with the MCRTreeCopier did not yet commit its OCFL Transaction, so the files which should be tiled do not exist yet.
We introduced Transaction commits in the MCRTreeCopier, so that in a large upload the connection to the database is released. In my opinion the cleanest aproach would be to remove the transaction restarts from the MCRTreeCopier, to always keep the file transaction and the database transaction in sync and also prevent weird rollback errors.

Meanwhile the uploads happen in another transaction and are stored in a local directory, so the "long" transaction only runs long to upload from the mycore machine to s3, or to copy from one local to another local directory, which should not take that long.

sebhofmann commented on mycore-ocfl/src/main/java/org/mycore/ocfl/metadata/MCROCFLXMLMetadataManagerAdapter.java:
```diff
@@ -338,23 +344,28 @@ public IntStream getStoredIDs(String project, String type) throws MCRPersistence
 
-        OcflExtensionConfig config = oclfRepoProvider.getExtensionConfig();
+        MCROCFLRepositoryProvider ocflRepoProvider = MCROCFLRepositoryProvider.getProvider(repositoryKey);
+        if (!(ocflRepoProvider instanceof MCROCFLLocalRepositoryProvider localRepositoryProvider)) {
```
This should still be possible, otherwise the user needs to rebuild the id cache by hand, which is a impossible task with the default layout.

Mewel replied:
I don't think it's a good idea to deactivate this check. I would rather have a migration command which builds the cache. Otherwise the remote repo is not useable anyway.

sebhofmann replied:
There is a command which builds the cache, but it uses this method. The method should not be called anywhere, except in the command. All other places should use the cache or the database.

Mewel replied:
Lets talk about this tomorrow, cause this method works directly with the Path API. It wants to call getRepositoryRoot(). This is not available on a remote S3 repository. There is no root path.

So either this method is wrongly implemented and shouldn't work with Path's or you would need to migrate to the MCRFileBaseCacheObjectIDGenerator first in order to use OCFL S3 repos.

Mewel committed:
MCR-3126 fix MCROCFLCachingSeekableByteChannel S3 buffer bug

Mewel committed:
MCR-3126 improve ocfl

* add restore derivate command
* add describe object command
* fix MCRVersionedPath#getParent does not return a versioned path

Mewel committed:
MCR-3126 fix rolling storage was not tracked properly

Mewel committed:
MCR-3126 rename some storage classes to be more precise

Mewel committed:
MCR-3126 prepare for rewriting rolling cache

Mewel committed:
MCR-3126 improve ocfl

- simplify virtual objects
- rewrote remote storage to use digest instead of path
- update junit test
- improve javadoc

Mewel committed:
MCR-3126 fix javadoc

Mewel committed:
MCR-3126 support attribute view in FileStore

Mewel committed:
MCR-3126 improve remote storage

- cp from transactional storage to remote storage after commit
- use "real" CAS files in remote storage

Mewel committed:
MCR-3126 fix cp to remote storage

the first commit to a derivate was not stored on the remote storage due to wrongful calculation of the version path

Mewel committed:
MCR-3126 stream directly to target

in remote repos stream directly to target path when makeing a
copy/externalCopy instead of streaming the source first and then copy
from there

Mewel committed:
MCR-3126 use new normalize() for digest in journal

Mewel committed:
MCR-3126 simplify remote storage journal

Mewel committed:
MCR-3126 rm TOUCH from journal

- makes the journal smaller
- compact every hour to store LRU queue
- memory is enough to store "touch" position

Mewel committed:
MCR-3126 small remote storage improvements

Mewel commented on mycore-ocfl/src/test/java/org/mycore/ocfl/metadata/migration/MCROCFLCombineIgnoreXPathPrunerTest.java:
```diff
@@ -28,16 +31,22 @@
-import org.junit.Assert;
-import org.junit.Test;
-import org.mycore.common.MCRTestCase;
+import org.junit.jupiter.api.Test;
```
This is incorrect.

Mewel committed:
MCR-3126 fix copilot requests

Mewel committed:
MCR-3126 use "old" highest id generation for remote repositories

Mewel committed:
MCR-3126 rm unused LOGGER

Mewel committed:
MCR-3126 use expanded object

Mewel committed:
MCR-3126 fix pom.xml

toKrause commented on mycore-ocfl/src/main/java/org/mycore/ocfl/repository/MCROCFLRepositoryProvider.java:
```diff
@@ -19,24 +19,66 @@
+     * @return the {@link MCROCFLRepositoryProvider} instance for the specified ID.
+     * @throws MCRConfigurationException if no provider is configured for the specified ID.
+     */
+    static MCROCFLRepositoryProvider getProvider(String id) {
```
Name `obtainInstance` would be more in line with the rest of MyCoRe.

toKrause commented on mycore-ocfl/pom.xml:
```diff
@@ -63,11 +87,48 @@
       <artifactId>mycore-user2</artifactId>
     </dependency>
+    <dependency>
+      <groupId>software.amazon.awssdk</groupId>
```
Presumably, the dependencies for S3-support could be marked as optional. 

Not everyone who wishes to use OCFL is going to also use S3. Theese dependecies could therefore be keept off the classpath for, presumably, a lot of users.

As far as I  can see, S3-supoport is implementes soley in `MCROCFLS3RepositoryProvider`.
Have you consideres moving S3-support into a separate module `mycore-ocfl-s3` that contains only that class, the corresponding entries in `mycore.properties` and the corresponding dependencies in `pom.xml`.

Not a "must", but would be nice.

Mewel committed:
MCR-3126 rename to MCROCFLRepositoryProvider#obtainInstance

Mewel committed:
MCR-3126 add mycore-ocfl-s3 module

Mewel committed:
MCR-3126 add missing mycore-ocfl-s3 to dependencyManagement
